runtime/near-vm-logic/src/receipt_manager.rs

```diff
@@ -0,0 +1,493 @@
+                predecessor_id: predecessor_id.clone(),
+                receiver_id,
+                // Actual receipt ID is set in the Runtime.apply_action_receipt(...) in the
+                // "Generating receipt IDs" section
+                receipt_id: CryptoHash::default(),
+                receipt: ReceiptEnum::Action(ActionReceipt {
+                    signer_id: signer_id.clone(),
+                    signer_public_key: signer_public_key.clone(),
+                    gas_price,
+                    output_data_receivers: receipt.output_data_receivers,
+                    input_data_ids: receipt.input_data_ids,
+                    actions: receipt.actions,
+                }),
+            })
+            .collect()
+    }
+}
+
+#[derive(Default, Clone, PartialEq)]
+pub(crate) struct ReceiptManager {
+    pub(crate) action_receipts: ActionReceipts,
+    #[cfg(feature = "protocol_feature_function_call_weight")]
+    gas_weights: Vec<(FunctionCallActionIndex, GasWeight)>,
+}
+
+#[cfg(feature = "protocol_feature_function_call_weight")]
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct FunctionCallActionIndex {
+    receipt_index: usize,
+    action_index: usize,
+}
+
+impl ReceiptManager {
+    pub fn get_receipt_receiver(&self, receipt_index: u64) -> Option<&AccountId> {
+        self.action_receipts.0.get(receipt_index as usize).map(|(id, _)| id)
+    }
+
+    /// Appends an action and returns the index the action was inserted in the receipt
+    pub fn append_action(&mut self, receipt_index: u64, action: Action) -> usize {
+        let actions = &mut self
+            .action_receipts
+            .0
+            .get_mut(receipt_index as usize)
+            .expect("receipt index should be present")
+            .1
+            .actions;
+
+        actions.push(action);
+
+        // Return index that action was inserted at
+        actions.len() - 1
+    }
+
+    /// Create a receipt which will be executed after all the receipts identified by
+    /// `receipt_indices` are complete.
+    ///
+    /// If any of the [`RecepitIndex`]es do not refer to a known receipt, this function will fail
+    /// with an error.
+    ///
+    /// # Arguments
+    ///
+    /// * `generate_data_id` - function to generate a data id to connect receipt output to
+    /// * `receipt_indices` - a list of receipt indices the new receipt is depend on
+    /// * `receiver_id` - account id of the receiver of the receipt created
+    pub fn create_receipt(
+        &mut self,
+        ext: &mut dyn External,
+        receipt_indices: Vec<u64>,
+        receiver_id: AccountId,
+    ) -> ExtResult<u64> {
+        let mut input_data_ids = vec![];
+        for receipt_index in receipt_indices {
+            let data_id = ext.generate_data_id();
+            self.action_receipts
+                .0
+                .get_mut(receipt_index as usize)
+                .ok_or_else(|| HostError::InvalidReceiptIndex { receipt_index })?
```

Comment:
:thinking: this is pre-existing, but it actually feels like it should be an assert. If/when we refactor ReceiptIndex to be a proper newtype index, rather than just a type alias, it'll be more clear.

Reply:
No action required in this PR!